x/signal/keeper.go

```diff
@@ -171,23 +201,39 @@ func (k Keeper) GetVotingPowerThreshold(ctx sdk.Context) sdkmath.Int {
 	return thresholdFraction.MulInt(totalVotingPower).Ceil().TruncateInt()
 }
 
-// ShouldUpgrade returns true if the signalling mechanism has concluded
-// that the network is ready to upgrade. It also returns the version
-// that the node should upgrade to.
-func (k *Keeper) ShouldUpgrade() (bool, uint64) {
-	return k.quorumVersion != 0, k.quorumVersion
+// ShouldUpgrade returns whether the signalling mechanism has concluded that the
+// network is ready to upgrade and the version to upgrade to. It returns false
+// and 0 if no version has reached quorum.
+func (k *Keeper) ShouldUpgrade(ctx sdk.Context) (isQuorumVersion bool, version uint64) {
+	upgrade, ok := k.getUpgrade(ctx)
+	if !ok {
+		return false, 0
+	}
+
+	hasUpgradeHeightBeenReached := ctx.BlockHeight() >= upgrade.UpgradeHeight
+	if hasUpgradeHeightBeenReached {
+		return true, upgrade.AppVersion
+	}
+	return false, 0
 }
 
 // ResetTally resets the tally after a version change. It iterates over the
-// store and deletes all versions. It also resets the quorumVersion to 0.
+// store and deletes all versions. It also resets the quorumVersion and
+// upgradeHeight to 0.
 func (k *Keeper) ResetTally(ctx sdk.Context) {
 	store := ctx.KVStore(k.storeKey)
 	iterator := store.Iterator(nil, nil)
 	defer iterator.Close()
+	// delete all signals
 	for ; iterator.Valid(); iterator.Next() {
+		if bytes.Equal(iterator.Key(), types.UpgradeKey) {
+			// skip over the upgrade key
+			continue
+		}
 		store.Delete(iterator.Key())
 	}
-	k.quorumVersion = 0
+	// delete the upgrade value
+	store.Delete(types.UpgradeKey)
```

Comment:
Is it necessary to delete the upgrade last instead of first?

Reply:
nope but I thought this was clearer. I refactored to something else that may be clearer: c7a9f7d43646cb168750d4709ea93646be226b3c